core/src/test/scala/shapeless/generic.scala

```diff
@@ -265,6 +272,17 @@ class GenericTests {
     typed[AbstractSingle](s1)
   }
 
+  @Test
+  def testOverlappingCoproducts {
+    val gen = Generic[Overlapping]
+    val o: Overlapping = OAB(1)
+    val o0 = gen.to(o)
+    typed[OAB :+: OAC :+: OBC :+: CNil](o0)
```

Comment:
I think it's a bit weird to auto-flatten the type hierarchy like this, wouldn't it make sense to infer `type Repr = OA :+: OB :+: CNil` instead?

Reply:
I could see a case being made for that.